scheduler/src/cook/mesos/unscheduled.clj

```diff
@@ -0,0 +1,157 @@
+  each type of failure}].
+  Otherwise, if job wasn't already under investigation,
+  returns [String indicating job is now under investigation, {}] "
+  [conn job]
+  (when-not (:job/under-investigation job)
+    @(d/transact conn [{:db/id (:db/id job)
+                        :job/under-investigation true}]))
+  (if (:job/last-fenzo-placement-failure job)
+    ["The job couldn't be placed on any available hosts."
+     {:reasons (-> job :job/last-fenzo-placement-failure edn/read-string
+                   fenzo-failures-for-user)}]
+    ["The job is now under investigation. Check back in a minute for more details!" {}]))
+
+(defn check-queue-position
+  "IFF the job is not first in the user's queue, returns
+  [\"You have x other jobs ahead in the queue\", {:jobs [other job uuids]]}]"
+  [conn job]
+  (let [db (d/db conn)
+        user (:job/user job)
+        job-uuid (:job/uuid job)
+        running-tasks (map
+                       (fn [j] (->> j
+                                    :job/instance
+                                    (filter util/instance-running?)
+                                    last))
+                       (util/jobs-by-user-and-state db user :job.state/running))
+        pending-tasks (map util/create-task-ent
+                           (util/jobs-by-user-and-state db user :job.state/waiting))
+        all-tasks (into running-tasks pending-tasks)
+        sorted-tasks (vec (sort (util/same-user-task-comparator) all-tasks))
+        queue-pos (first
+                   (keep-indexed
+                    (fn [i instance]
+                      (when (= (-> instance :job/_instance :job/uuid) job-uuid) i))
+                    sorted-tasks))
+        tasks-ahead (subvec sorted-tasks 0 queue-pos)]
+    (when (seq tasks-ahead)
+      [(str "You have " queue-pos " other jobs ahead in the queue.")
+       {:jobs (mapv #(-> % :job/_instance :job/uuid str) tasks-ahead)}])))
+
+(defn reasons
+  "Top level function which assembles a data structure representing the list
+  of possible responses to the question \"Why isn't this job being scheduled?\".
+  This is a vector of tuples of (answer, extra data about answer)
+  e.g. [[\"Reason One\" {:key1 1 :key2 2}]
+        [\"Reason Two\" {:keya 3 :keyb 4}]]
+  Possible responses can include essentially saying,
+  \"Actually, it WAS scheduled\"\" e.g. [[\"The job is running now.\" {}]]."
+  [conn job]
+  (let [db (d/db conn)]
+    (case (:job/state job)
+      :job.state/running [["The job is running now." {}]]
+      :job.state/completed [["The job already completed." {}]]
+      (filter some?
+              [(check-exhausted-retries job)
+               (check-exceeds-limit quota/get-quota
+                                    "The job would cause you to exceed resource quotas."
+                                    db job)
+               (check-exceeds-limit share/get-share
+                                    "The job would cause you to exceed resource shares."
```

Comment:
It is possible to be scheduled even if your job exceeds the resource share. I'm not sure the best way to address this.